File: hmh-frontend/src/api/workshop.ts
import client from "./client";

export type MRPriority = "LOW" | "NORMAL" | "HIGH" | "URGENT";
export type WorkshopMRStatus = "DRAFT" | "SUBMITTED" | "APPROVED" | "REJECTED";

export interface WorkshopCategory {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface WorkshopItem {
  id: string;
  category_id: string;
  name: string;
  part_number: string | null;
  unit: string;
  description: string | null;
  reorder_level: number | null;
  is_active: boolean;
  quantity_on_hand: number;
  created_at: string;
  updated_at: string;
}

export interface WorkshopMRLineBrief {
  id: string;
  workshop_mr_id: string;
  item_id: string;
  quantity_requested: number;
  quantity_approved: number | null;
  preferred_supplier_id: string | null;
  remarks: string | null;
  item: { id: string; name: string; unit: string; part_number: string | null } | null;
}

export interface WorkshopMRApproval {
  id: string;
  mr_id: string;
  approved_by: string | null;
  approved_at: string;
  is_override: boolean;
  notes: string | null;
  voter: { id: string; full_name: string; email: string } | null;
}

export interface WorkshopMR {
  id: string;
  mr_number: string;
  site_id: string;
  vehicle_id: string;
  reason: string;
  status: WorkshopMRStatus;
  priority: MRPriority;
  needed_by_date: string | null;
  requested_by: string;
  approved_by: string | null;
  approved_at: string | null;
  rejection_reason: string | null;
  notes: string | null;
  created_at: string;
  updated_at: string;
  lines: WorkshopMRLineBrief[];
  approvals: WorkshopMRApproval[];
  vote_count: number;
  // Resolved names (embedded by backend)
  site: { id: string; name: string; site_type: string } | null;
  vehicle: { id: string; registration: string; name: string } | null;
}

export interface WorkshopIssuance {
  id: string;
  item_id: string;
  vehicle_id: string;
  workshop_mr_id: string | null;
  quantity_issued: number;
  issued_by: string;
  issued_at: string;
  notes: string | null;
  created_at: string;
  item: { id: string; name: string; unit: string; part_number: string | null } | null;
  vehicle: { id: string; registration: string; name: string } | null;
}

export interface WorkshopIssuanceCreate {
  item_id: string;
  vehicle_id: string;
  workshop_mr_id?: string | null;
  quantity_issued: number;
  notes?: string | null;
}

export interface WorkshopSupplierLink {
  id: string;
  category_id: string;
  supplier_id: string;
  is_preferred: boolean;
  category: { id: string; name: string; description: string | null } | null;
  supplier: { id: string; name: string; email: string | null } | null;
}

export interface WorkshopCategoryCreate {
  name: string;
  description?: string | null;
}

export interface WorkshopItemCreate {
  category_id: string;
  name: string;
  part_number?: string | null;
  unit?: string;
  description?: string | null;
  reorder_level?: number | null;
}

export interface WorkshopItemUpdate {
  name?: string;
  part_number?: string | null;
  unit?: string;
  description?: string | null;
  reorder_level?: number | null;
  is_active?: boolean;
}

export interface WorkshopSupplierLinkCreate {
  category_id: string;
  supplier_id: string;
  is_preferred?: boolean;
}

export interface WorkshopMRLineCreate {
  item_id: string;
  quantity_requested: number;
  preferred_supplier_id?: string;
  remarks?: string;
}

export interface WorkshopMRCreate {
  site_id: string;
  vehicle_id: string;
  reason: string;
  priority?: MRPriority;
  needed_by_date?: string | null;
  notes?: string | null;
  lines: WorkshopMRLineCreate[];
}

export const workshopApi = {
  listCategories: async (): Promise<WorkshopCategory[]> => {
    const res = await client.get<{ data: WorkshopCategory[] }>("/workshop/categories/");
    return res.data.data;
  },

  listItems: async (categoryId?: string): Promise<WorkshopItem[]> => {
    const params: Record<string, string> = { active_only: "true" };
    if (categoryId) params.category_id = categoryId;
    const res = await client.get<{ data: WorkshopItem[] }>("/workshop/items/", { params });
    return res.data.data;
  },

  listMRs: async (opts?: {
    site_id?: string;
    vehicle_id?: string;
    status?: WorkshopMRStatus;
  }): Promise<WorkshopMR[]> => {
    const params: Record<string, string> = {};
    if (opts?.site_id)    params.site_id    = opts.site_id;
    if (opts?.vehicle_id) params.vehicle_id = opts.vehicle_id;
    if (opts?.status)     params.status     = opts.status;
    const res = await client.get<{ data: WorkshopMR[] }>("/workshop/mrs/", { params });
    return res.data.data;
  },

  getMR: async (mrId: string): Promise<WorkshopMR> => {
    const res = await client.get<{ data: WorkshopMR }>(`/workshop/mrs/${mrId}`);
    return res.data.data;
  },

  createMR: async (body: WorkshopMRCreate): Promise<WorkshopMR> => {
    const res = await client.post<{ data: WorkshopMR }>("/workshop/mrs/", body);
    return res.data.data;
  },

  submitMR: async (mrId: string): Promise<WorkshopMR> => {
    const res = await client.post<{ data: WorkshopMR }>(`/workshop/mrs/${mrId}/submit`);
    return res.data.data;
  },

  approveMR: async (mrId: string): Promise<WorkshopMR> => {
    const res = await client.post<{ data: WorkshopMR }>(`/workshop/mrs/${mrId}/approve`);
    return res.data.data;
  },

  rejectMR: async (mrId: string, reason?: string): Promise<WorkshopMR> => {
    const res = await client.post<{ data: WorkshopMR }>(`/workshop/mrs/${mrId}/reject`, {
      reason: reason ?? null,
    });
    return res.data.data;
  },

  castVote: async (mrId: string, notes?: string): Promise<WorkshopMR> => {
    const res = await client.post<{ data: WorkshopMR }>(`/workshop/mrs/${mrId}/vote`, {
      notes: notes ?? null,
    });
    return res.data.data;
  },

  // ── Categories ────────────────────────────────────────────────────────────

  createCategory: async (body: WorkshopCategoryCreate): Promise<WorkshopCategory> => {
    const res = await client.post<{ data: WorkshopCategory }>("/workshop/categories/", body);
    return res.data.data;
  },

  updateCategory: async (categoryId: string, body: WorkshopCategoryCreate): Promise<WorkshopCategory> => {
    const res = await client.patch<{ data: WorkshopCategory }>(`/workshop/categories/${categoryId}`, body);
    return res.data.data;
  },

  // ── Items ─────────────────────────────────────────────────────────────────

  createItem: async (body: WorkshopItemCreate): Promise<WorkshopItem> => {
    const res = await client.post<{ data: WorkshopItem }>("/workshop/items/", body);
    return res.data.data;
  },

  updateItem: async (itemId: string, body: WorkshopItemUpdate): Promise<WorkshopItem> => {
    const res = await client.patch<{ data: WorkshopItem }>(`/workshop/items/${itemId}`, body);
    return res.data.data;
  },

  adjustStock: async (itemId: string, quantityDelta: number): Promise<{ quantity_on_hand: number }> => {
    const res = await client.post<{ data: { quantity_on_hand: number } }>(
      `/workshop/items/${itemId}/adjust-stock`,
      { quantity_delta: quantityDelta },
    );
    return res.data.data;
  },

  // ── Supplier links ────────────────────────────────────────────────────────

  listSupplierLinks: async (categoryId?: string): Promise<WorkshopSupplierLink[]> => {
    const params: Record<string, string> = {};
    if (categoryId) params.category_id = categoryId;
    const res = await client.get<{ data: WorkshopSupplierLink[] }>("/workshop/supplier-links/", { params });
    return res.data.data;
  },

  createSupplierLink: async (body: WorkshopSupplierLinkCreate): Promise<WorkshopSupplierLink> => {
    const res = await client.post<{ data: WorkshopSupplierLink }>("/workshop/supplier-links/", body);
    return res.data.data;
  },

  deleteSupplierLink: async (linkId: string): Promise<void> => {
    await client.delete(`/workshop/supplier-links/${linkId}`);
  },

  // ── Issuances ─────────────────────────────────────────────────────────────

  listIssuances: async (opts?: { vehicle_id?: string; item_id?: string }): Promise<WorkshopIssuance[]> => {
    const params: Record<string, string> = {};
    if (opts?.vehicle_id) params.vehicle_id = opts.vehicle_id;
    if (opts?.item_id)    params.item_id    = opts.item_id;
    const res = await client.get<{ data: WorkshopIssuance[] }>("/workshop/issuances/", { params });
    return res.data.data;
  },

  issueParts: async (body: WorkshopIssuanceCreate): Promise<WorkshopIssuance> => {
    const res = await client.post<{ data: WorkshopIssuance }>("/workshop/issuances/", body);
    return res.data.data;
  },
};

File: hmh-backend/app/schemas/workshop.py
"""Pydantic v2 schemas for the Workshop module."""

import uuid
from datetime import date, datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, computed_field, field_validator

from app.models.enums import MRPriority, RecordStatus


# ── Stock ─────────────────────────────────────────────────────────────────────

class WorkshopStockRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    quantity_on_hand: float = 0.0
    last_updated: Optional[datetime] = None


# ── Categories ────────────────────────────────────────────────────────────────

class WorkshopCategoryRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    description: Optional[str] = None
    created_at: datetime
    updated_at: datetime


# ── Embedded resolved references ──────────────────────────────────────────────

class _SiteBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    site_type: str


class _VehicleBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    registration: str
    name: str


class _UserBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    full_name: str
    email: str


class WorkshopCategoryCreate(BaseModel):
    name: str
    description: Optional[str] = None

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Category name is required")
        return v.strip()


# ── Items ─────────────────────────────────────────────────────────────────────

class WorkshopItemRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    category_id: uuid.UUID
    name: str
    part_number: Optional[str] = None
    unit: str
    description: Optional[str] = None
    reorder_level: Optional[float] = None
    is_active: bool
    stock: Optional[WorkshopStockRead] = None
    category: Optional[WorkshopCategoryRead] = None
    created_at: datetime
    updated_at: datetime

    @computed_field
    @property
    def quantity_on_hand(self) -> float:
        return float(self.stock.quantity_on_hand) if self.stock else 0.0

    @computed_field
    @property
    def category_name(self) -> Optional[str]:
        return self.category.name if self.category else None


class WorkshopItemCreate(BaseModel):
    category_id: uuid.UUID
    name: str
    part_number: Optional[str] = None
    unit: str = "each"
    description: Optional[str] = None
    reorder_level: Optional[float] = None

    @field_validator("name")
    @classmethod
    def name_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Item name is required")
        return v.strip()


class WorkshopItemUpdate(BaseModel):
    name: Optional[str] = None
    part_number: Optional[str] = None
    unit: Optional[str] = None
    description: Optional[str] = None
    reorder_level: Optional[float] = None
    is_active: Optional[bool] = None


# ── Supplier links ─────────────────────────────────────────────────────────────

class _SupplierBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    email: Optional[str] = None


class WorkshopSupplierLinkRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    category_id: uuid.UUID
    supplier_id: uuid.UUID
    is_preferred: bool
    category: Optional[WorkshopCategoryRead] = None
    supplier: Optional[_SupplierBrief] = None


class WorkshopSupplierLinkCreate(BaseModel):
    category_id: uuid.UUID
    supplier_id: uuid.UUID
    is_preferred: bool = False


# ── MR lines ──────────────────────────────────────────────────────────────────

class WorkshopItemBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    unit: str
    part_number: Optional[str] = None


class WorkshopMRLineRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    item_id: uuid.UUID
    quantity_requested: float
    quantity_approved: Optional[float] = None
    preferred_supplier_id: Optional[uuid.UUID] = None
    remarks: Optional[str] = None
    item: Optional[WorkshopItemBrief] = None


class WorkshopMRLineCreate(BaseModel):
    item_id: uuid.UUID
    quantity_requested: float
    preferred_supplier_id: Optional[uuid.UUID] = None
    remarks: Optional[str] = None

    @field_validator("quantity_requested")
    @classmethod
    def qty_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("Quantity must be greater than zero")
        return v


# ── MR Email Logs ────────────────────────────────────────────────────────────

class _SupplierEmailBrief(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    email: Optional[str] = None


class WorkshopMREmailLogRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    supplier_id: Optional[uuid.UUID] = None
    sent_to_email: str
    email_subject: Optional[str] = None
    status: str
    error_message: Optional[str] = None
    sent_at: Optional[datetime] = None
    created_at: datetime
    supplier: Optional[_SupplierEmailBrief] = None


# ── MR Approvals ─────────────────────────────────────────────────────────────

class WorkshopMRApprovalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    mr_id: uuid.UUID
    approved_by: Optional[uuid.UUID] = None
    approved_at: datetime
    is_override: bool
    notes: Optional[str] = None
    voter: Optional[_UserBrief] = None


# ── Workshop MRs ──────────────────────────────────────────────────────────────

class WorkshopMRRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    mr_number: str
    site_id: uuid.UUID
    vehicle_id: uuid.UUID
    reason: str
    status: RecordStatus
    priority: MRPriority
    needed_by_date: Optional[date] = None
    requested_by: uuid.UUID
    approved_by: Optional[uuid.UUID] = None
    approved_at: Optional[datetime] = None
    rejection_reason: Optional[str] = None
    notes: Optional[str] = None
    created_at: datetime
    updated_at: datetime
    lines: list[WorkshopMRLineRead] = []
    approvals: list[WorkshopMRApprovalRead] = []
    email_logs: list[WorkshopMREmailLogRead] = []
    quotes: list["WorkshopQuoteRead"] = []
    purchase_orders: list["WorkshopPurchaseOrderRead"] = []
    invoices: list["WorkshopInvoiceRead"] = []
    delivery_notes: list["WorkshopDeliveryNoteRead"] = []
    site: Optional[_SiteBrief] = None
    vehicle: Optional[_VehicleBrief] = None

    @computed_field
    @property
    def vote_count(self) -> int:
        return sum(1 for a in self.approvals if not a.is_override)


class WorkshopMRCreate(BaseModel):
    site_id: uuid.UUID
    vehicle_id: uuid.UUID
    reason: str
    priority: MRPriority = MRPriority.NORMAL
    needed_by_date: Optional[date] = None
    notes: Optional[str] = None
    lines: list[WorkshopMRLineCreate] = []

    @field_validator("reason")
    @classmethod
    def reason_not_empty(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Reason is required")
        return v.strip()

    @field_validator("lines")
    @classmethod
    def at_least_one_line(cls, v: list) -> list:
        if not v:
            raise ValueError("At least one line item is required")
        return v


# ── Workshop Quotes ───────────────────────────────────────────────────────────

class WorkshopQuoteApprovalRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    quote_id: uuid.UUID
    approved_by: Optional[uuid.UUID] = None
    approved_at: datetime
    is_override: bool
    notes: Optional[str] = None
    voter: Optional[_UserBrief] = None


class WorkshopQuoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    quote_file_url: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str
    notes: Optional[str] = None
    status: str
    rejection_reason: Optional[str] = None
    submitted_at: Optional[datetime] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    approvals: list[WorkshopQuoteApprovalRead] = []
    supplier: Optional[_SupplierBrief] = None

    @computed_field
    @property
    def vote_count(self) -> int:
        return sum(1 for a in self.approvals if not a.is_override)


class WorkshopQuoteCreate(BaseModel):
    workshop_mr_id: uuid.UUID
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str = "ZAR"
    notes: Optional[str] = None


# ── Purchase Orders ───────────────────────────────────────────────────────────

class WorkshopPurchaseOrderRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    quote_id: Optional[uuid.UUID] = None
    po_number: str
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str
    notes: Optional[str] = None
    status: str
    po_file_url: Optional[str] = None
    sent_at: Optional[datetime] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    supplier: Optional[_SupplierBrief] = None


class WorkshopPurchaseOrderCreate(BaseModel):
    workshop_mr_id: uuid.UUID
    quote_id: Optional[uuid.UUID] = None
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str = "ZAR"
    notes: Optional[str] = None


# ── Invoices ──────────────────────────────────────────────────────────────────

class WorkshopInvoiceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    po_id: Optional[uuid.UUID] = None
    invoice_number: Optional[str] = None
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    invoice_file_url: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str
    invoice_date: Optional[date] = None
    notes: Optional[str] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime
    supplier: Optional[_SupplierBrief] = None


class WorkshopInvoiceCreate(BaseModel):
    workshop_mr_id: uuid.UUID
    po_id: Optional[uuid.UUID] = None
    invoice_number: Optional[str] = None
    supplier_id: Optional[uuid.UUID] = None
    supplier_name: Optional[str] = None
    total_amount: Optional[float] = None
    currency: str = "ZAR"
    invoice_date: Optional[date] = None
    notes: Optional[str] = None


# ── Delivery Notes ────────────────────────────────────────────────────────────

class WorkshopDeliveryNoteRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    workshop_mr_id: uuid.UUID
    po_id: Optional[uuid.UUID] = None
    delivery_number: Optional[str] = None
    delivery_file_url: Optional[str] = None
    delivery_date: Optional[date] = None
    received_by_name: Optional[str] = None
    notes: Optional[str] = None
    created_by: Optional[uuid.UUID] = None
    created_at: datetime
    updated_at: datetime


class WorkshopDeliveryNoteCreate(BaseModel):
    workshop_mr_id: uuid.UUID
    po_id: Optional[uuid.UUID] = None
    delivery_number: Optional[str] = None
    delivery_date: Optional[date] = None
    received_by_name: Optional[str] = None
    notes: Optional[str] = None


# Update forward refs
WorkshopMRRead.model_rebuild()


# ── Issuances ─────────────────────────────────────────────────────────────────

class WorkshopIssuanceRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    item_id: uuid.UUID
    vehicle_id: uuid.UUID
    workshop_mr_id: Optional[uuid.UUID] = None
    quantity_issued: float
    issued_by: uuid.UUID
    issued_at: datetime
    notes: Optional[str] = None
    created_at: datetime
    item:    Optional[WorkshopItemBrief] = None
    vehicle: Optional[_VehicleBrief]    = None


class WorkshopIssuanceCreate(BaseModel):
    item_id: uuid.UUID
    vehicle_id: uuid.UUID
    workshop_mr_id: Optional[uuid.UUID] = None
    quantity_issued: float
    notes: Optional[str] = None

    @field_validator("quantity_issued")
    @classmethod
    def qty_positive(cls, v: float) -> float:
        if v <= 0:
            raise ValueError("Quantity must be greater than zero")
        return v
